### ISSUE TITLE
Clarification on KL Divergence Computation in GRPOTrainer

### ISSUE THREAD

**Issue by zhaopku** [huggingface/trl#2914]
Hi team,

I'm trying to understand the exact form of the KL divergence computed in the GRPOTrainer's `compute_loss()` function in the Hugging Face TRL repository. The code computes the per-token KL divergence as follows:

```python
per_token_kl = torch.exp(ref_per_token_logps - per_token_logps) - (ref_per_token_logps - per_token_logps) - 1
```

From my understanding, the KL divergence can be defined in two ways:

- **Forward KL:**  
$`
  D_{\mathrm{KL}}(p_{\mathrm{ref}} \,\|\, p_{\mathrm{current}}) = \mathbb{E}_{x \sim p_{\mathrm{ref}}}\left[\log \frac{p_{\mathrm{ref}}(x)}{p_{\mathrm{current}}(x)}\right]
`$

- **Reverse KL:**  
$`
  D_{\mathrm{KL}}(p_{\mathrm{current}} \,\|\, p_{\mathrm{ref}}) = \mathbb{E}_{x \sim p_{\mathrm{current}}}\left[\log \frac{p_{\mathrm{current}}(x)}{p_{\mathrm{ref}}(x)}\right]
`$

In the code, you define:

$`\Delta_t = log(p_{ref}(x_t)) - log(p_{current}(x_t))`$

and then compute:

$`KL_t = exp(\Delta_t) - \Delta_t - 1`$

Since tokens are sampled from the current policy  $\(p_{\mathrm{current}}\)$, I would expect a reverse KL divergence. However, the code looks like what I'd typically associate with forward KL.

Could you please clarify:
1. Which KL divergence is the code intended to compute? Is it closer to the forward KL or the reverse KL?
2. What is the rationale behind using the function $\exp(\Delta_t) - \Delta_t - 1$ as a KL penalty in this context?

Thank you for your help in clarifying this!

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
Thank for the question :) The answer is in the paper:

<img width="1114" alt="Image" src="https://github.com/user-attachments/assets/c8278f9c-3d65-4e05-97c8-8a9bcd35073b" />

The ref is http://joschu.net/blog/kl-approx.html

**Comment by aburkov** [comment]
In the paper, there's Algorithm 1 looking this way:

![Image](https://github.com/user-attachments/assets/57a3eedb-b694-4f78-9669-fdbf83bdbb95)

It makes reference to Equation 21 they call gradient coeficient:

![Image](https://github.com/user-attachments/assets/f4613cb5-887f-4817-acaa-1f6da5d2a9a2)

Is there a reason why it wasn't implemented as Algorithm 1 defines it?

**Comment by qgallouedec** [comment]
answered here https://github.com/huggingface/trl/issues/2752

**Comment by aburkov** [comment]
> answered here [#2752](https://github.com/huggingface/trl/issues/2752)

Are you referring to this answer:

> Equation (21) gives the gradient of the objective, but what we're trying to optimize is the objective, not the gradient. This is why we use equation (3) in practice, and autograd takes care of the derivation and optimization step.

? If yes, I understand that. My question was why didn't you use Algorithm 1 and, instead, decided to implement a more complex loss that PyTorch then must autograd, if you already have a simplified expression for the gradient in Equation 21 and Algorithm 1 showing how to use it?

**Comment by qgallouedec** [comment]
How would you do this?

**Comment by aburkov** [comment]
> How would you do this?

That's why I'm asking. I implemented it as Algorithm 1 with Equation 20 as the loss, the way the paper describes it, but it doesn't learn. So, I thought that maybe I misunderstood something in Algorithm 1 and tried to find a different implementation, but all implementations I found are using the same objective you used, which is not what Algorithm 1 says should be used. So I asked you about the reasons for you not to implement it as Algorithm 1. Is it an error in the paper? How did you come to the conclusion that the loss shouldn't be the equations 20/21?

**Comment by qgallouedec** [comment]
equation 21 is the derivative of the loss, not the loss, see https://github.com/huggingface/trl/issues/2752#issuecomment-2635204212. So either you use the loss as expressed in equation (3), what we usually do, or you directly use it's derivative by _disabling_ autograd (not sure why and how you would do that).

**Comment by aburkov** [comment]
Algorithm 1 says "by maximizing the GRPO objective, Equation 21." If you try to maximize equation 21, the model doesn't learn. I'm not sure if I clearly ask the question. How the article reader wanting to implement GRPO is supposed to figure out that Equation 21 referred as the objective in Algorithm 1 is actually equation 3?

**Comment by qgallouedec** [comment]
I suppose it could have been expressed more clearly, but we never maximize the “gradient coefficient”. It doesn't really make sense. We maximize the objective using the gradient. That's how I understood it. If you want more information on why it is phrased this way you'd have to contact the author directly.

**Comment by aburkov** [comment]
No, thanks. I'm sure the way you understood it is the right one. I tested with both equations 20 and 21 as objectives, and it doesn't learn. So, testing equation 3 is the next logical choice :-) They just didn't reference the right equation in Algorithm 1.
